File: solutions/ruby/collatz-conjecture/6/collatz_conjecture.rb
class CollatzConjecture

  class InvalidNumberError < ArgumentError
    def initialize message = 'Number must be a positive integer'
      super message
    end
  end

  INFINITY = 1.step
  RULE = 'number.even? ? number / 2 : number * 3 + 1'
  COLLATZ = proc { |number| eval RULE }
  private_constant :INFINITY, :RULE, :COLLATZ

  def self.steps number
    new(number).steps
  end

  private

  attr_writer :number, :steps

  def initialize number
    number.positive? or raise InvalidNumberError
    self.number = number
    self.steps = calculate_steps
  end

  def calculate_steps
    current = number

    INFINITY.count do |step|
      return step - 1 if current == 1
      current = COLLATZ[current]
    end
  end

  public

  attr_accessor :number, :steps

end

File: solutions/elixir/collatz-conjecture/6/lib/collatz_conjecture.ex
defmodule CollatzConjecture do

  def calc(number) when is_integer(number) and number >= 1,
    do: steps(number, 0)

  defp steps(1, steps), do: steps

  defp steps(n, steps) when rem(n, 2) == 0,
    do: steps(div(n, 2), steps + 1)

  defp steps(n, steps),
    do: steps(3 * n + 1, steps + 1)

end
